mindmeld/active_learning/alp.py

```diff
@@ -0,0 +1,234 @@
+        logger.info("Creating Training Data Bucket")
+        self.output_manager.create_experiment_folder(self.training_strategies)
+        self.init_data_bucket = DataBucketFactory.get_data_bucket_for_training(
+            self.app_path,
+            self.load,
+            self.save,
+            self.train_pattern,
+            self.test_pattern,
+            self.train_seed_pct,
+        )
+        logger.info("Starting Training")
+        for strategy in self.training_strategies:
+            self._train_strategy(strategy)
+
+    def _train_strategy(
+        self,
+        strategy: str,
+        selection_mode: bool = None,
+    ):
+        """Helper function to traing a single strategy.
+
+        Args:
+            strategy (str): Single strategy to train
+            selection_mode (bool): If in selection mode, accuracies will not be recorded
+                and the run will terminate after the first iteration
+        """
+        early_stop = False
+        for epoch in range(self.n_epochs):
+            del self.data_bucket
+            self.data_bucket = deepcopy(self.init_data_bucket)
+            num_iterations = math.ceil(
+                len(self.data_bucket.unsampled_queries) / self.batch_size
+            )
+            iteration = 0
+            while iteration <= num_iterations:
+                logger.info(
+                    "Strategy: %s. Epoch: %s. Iteration %s.", strategy, epoch, iteration
+                )
+                logger.info(
+                    "Sampled Queries: %s", len(self.data_bucket.sampled_queries)
+                )
+                logger.info(
+                    "Remaining (Unsampled) Queries: %s",
+                    len(self.data_bucket.unsampled_queries),
+                )
+                (
+                    eval_stats,
+                    preds_single,
+                    preds_multi,
+                    domain_indices,
+                ) = self.mindmeld_classifier.train(
+                    data_bucket=self.data_bucket,
+                    return_preds_multi=STRATEGY_ABRIDGED[strategy]
+                    in ["ds", "ens", "kld"],
+                )
+                if not selection_mode:
+                    self.output_manager.update_accuracies_json(
+                        strategy, epoch, iteration, eval_stats
+                    )
+                    if self.save_sampled_queries:
+                        self.output_manager.update_selected_queries_json(
+                            strategy,
+                            epoch,
+                            iteration,
+                            self.data_bucket.newly_sampled_queries,
+                        )
+                    if self.early_stopping_window > 0:
+                        early_stop = self.output_manager.check_early_stopping(
+                            strategy, self.early_stopping_window
+                        )
+                num_unsampled = len(self.data_bucket.unsampled_queries)
```

Comment:
this can be outside the loop

Reply:
Since the number of "unsampled_queries" in the data_bucket will be updated on each iteration, it makes sense to recalculate num_unsampled within the loop